### PR TITLE
Enhancement Script Change (UEF-ACU)

### PR DESCRIPTION
**Pull request by MrRowey**
### Pull Request Summary

This Pull Request refactors the Enhancement handling logic in the script by replacing the long `if/else` chain with a modular approach where each enhancement has a dedicated function. This change aims to improve code maintainability and readability.

### Background

The current implementation stores enhancements in a lengthy `if/else` chain, making it difficult to manage and understand. After discussing four potential approaches for improvement in Discord, we concluded that the existing method (Option 1) is not viable due to the difficulty in maintaining and extending it. Here’s a summary of each option and the reasoning behind our choice:

1. **Current Implementation (if/else chain)**: Difficult to maintain and understand. Not viable.
2. **Enhancement Upgrades Table within Class with Key Lookup**: Improved maintainability but poses issues with inheritance and potential for modifications to leak through the hierarchy. This approach was ruled out due to concerns raised by hdt80: "Anything that requires several levels of nesting won't work well with inheritance and may confuse as to whether the table is part of the class structure."
3. **Enhancement Upgrades Table as a Module with Key Lookup**: While modular, this approach would require redefining the `CreateEnhancements` function with each new file, creating redundancy. This option was ruled out, as noted by Nomander: "You have to rewrite the enhancements table and `CreateEnhancements` every new file, which is inconvenient."
4. **Inline Enhancement Functions in Unit Class with Key Lookup**: Improves maintainability without complex inheritance issues. This approach aligns with standard methods for modding behaviour and allows each enhancement to be directly modified or extended.

Based on these discussions, **Option 4** was determined to be the most logical and maintainable choice.

### Implementation

With Option 4, each enhancement type (e.g., `Shield`, `ShieldRemove`) will now have its dedicated function. For instance, here’s how the `Shield` enhancement logic changes:

**Before (in `if/else` chain):**
```lua
elseif enh == 'Shield' then
    self:AddToggleCap('RULEUTC_ShieldToggle')
    self:CreateShield(bp)
    self:SetEnergyMaintenanceConsumptionOverride(bp.MaintenanceConsumptionPerSecondEnergy or 0)
    self:SetMaintenanceConsumptionActive()
elseif enh == 'ShieldRemove' then
    self:DestroyShield()
    self:SetMaintenanceConsumptionInactive()
    RemoveUnitEnhancement(self, 'ShieldRemove')
    self:RemoveToggleCap('RULEUTC_ShieldToggle')
```

**After (Option 4 with dedicated functions):**
```lua
ProcessShieldEnhancement = function(self, bp)
    self:AddToggleCap('RULEUTC_ShieldToggle')
    self:CreateShield(bp)
    self:SetEnergyMaintenanceConsumptionOverride(bp.MaintenanceConsumptionPerSecondEnergy or 0)
    self:SetMaintenanceConsumptionActive()
end,

ProcessShieldRemoveEnhancement = function(self, bp)
    self:DestroyShield()
    self:SetMaintenanceConsumptionInactive()
    RemoveUnitEnhancement(self, 'ShieldRemove')
    self:RemoveToggleCap('RULEUTC_ShieldToggle')
end,
```

### `CreateEnhancement` Function Redesign

The `CreateEnhancement` function now dynamically calls enhancement functions using a naming convention based on the enhancement name. For example, `Shield` enhancements will map to `ProcessShieldEnhancement`, making the code modular and error-resistant. The updated function is as follows:

```lua
CreateEnhancement = function(self, enh)
    ACUUnit.CreateEnhancement(self, enh)
    local bp = self:GetBlueprint().Enhancements[enh]
    if not bp then return end

    local ref = 'Process' .. enh .. 'Enhancement'
    local handler = self[ref]
    if handler then
        handler(self, bp)
    else
        WARN("Missing enhancement: ", enh, " for unit: ", self:GetUnitId(), " - expected function name: ", ref)
    end
end,
```

This approach offers:

- **Modularity**: Each enhancement is a self-contained function, simplifying future modifications.
- **Error Handling**: The new structure provides clear warnings if an enhancement function is missing, which improves troubleshooting.
- **Readability and Maintenance**: By removing the lengthy `if/else` chain, the code is easier to read, manage, and extend.

## Checklist
- [x] Changes are annotated, including comments where useful
- [x] Changes are documented in the changelog for the next game version
- [x] Tested that all Upgrades work in the game both Installing and removing each upgrade.

### CODE REVIEW

File: units/UEL0001/UEL0001_script.lua [hunk 1]
```diff
@@ -302,237 +302,388 @@ UEL0001 = ClassUnit(ACUUnit) {
...
+    ---@param self UEL0001
+    ---@param bp Blueprint unused
+    ProcessAdvancedEngineeringRemoveEnhancement = function(self, bp)
+        local bp = self:GetBlueprint().Economy.BuildRate
+        if not bp then return end
+        self:RestoreBuildRestrictions()
+        self:AddBuildRestriction(categories.UEF *
+            (categories.BUILTBYTIER2COMMANDER + categories.BUILTBYTIER3COMMANDER))
+        self:AddBuildRestriction(categories.UEF *
+            (categories.BUILTBYTIER2COMMANDER + categories.BUILTBYTIER3COMMANDER))
```
Comment (sprunk) [inline units/UEL0001/UEL0001_script.lua:482]: duplicate?

Comment (MrRowey) [reply units/UEL0001/UEL0001_script.lua:482]: All I've done is Coppy Past the old code within the if statement as you can see here

image

It been there all this time xD

So we're already improving maintainability as we can see like you says its a duplicate

Comment (sprunk) [reply units/UEL0001/UEL0001_script.lua:482]: Maybe it's good to keep it just a Coppy Past for now then (to ensure there are no behaviour changes) and do further improvements separately.

File: units/UEL0001/UEL0001_script.lua [hunk 2]
```diff
@@ -302,237 +302,388 @@ UEL0001 = ClassUnit(ACUUnit) {
...
+
+    ---@param self UEL0001
+    ---@param enh string
+    CreateEnhancement = function(self, enh)
+        ACUUnit.CreateEnhancement(self, enh)
+
+        local bp = self:GetBlueprint().Enhancements[enh]
+        if not bp then return end
+
+        local ref = 'Process' .. enh .. 'Enhancement'
```
Comment (Hdt80bro) [inline units/UEL0001/UEL0001_script.lua:681]: Consider renaming methods to `ProcessEnhancement`*name* so it's not hiding in between bookends like this

Comment (MrRowey) [reply units/UEL0001/UEL0001_script.lua]: Updated to use this better naming Schema

File: units/UEL0001/UEL0001_script.lua (hunk 1 above)
Comment (MrRowey) [reply units/UEL0001/UEL0001_script.lua:482]: Yup that would be the plan.

Comment (Garanas) [reply units/UEL0001/UEL0001_script.lua:482]: I agree, it is better to do that separate.

**Comment by Garanas** [comment]
Great job @MrRowey , looking forward to the other (S)ACUs 😃 .

And for small issues that @sprunk mentions, feel free to turn those into GitHub issues and assign the 'good-first-issue' label to it. They're great to introduce new contributors to the development environment.

**Pull request merged**